client/src/app/shared/containers/application-dependencies-form/application-dependencies-form.tsx

```diff
@@ -152,6 +153,20 @@ export const ApplicationDependenciesForm: React.FC<
           value={northboundDependencies}
           setValue={setNorthboundDependencies}
           options={(applications || [])
+            .filter((app) => {
+              let blacklistedCircularDeps: (number | undefined)[] = [];
+              if (southDependencies) {
+                southDependencies?.forEach((dep) => {
+                  if (dep.from.id === application.id) {
+                    let sbd = dep;
+                    blacklistedCircularDeps = southDependencies
+                      .filter((dep) => dep.from.id === sbd.to.id)
+                      .map((dep) => dep.to.id);
+                  }
+                });
+              }
+              return !blacklistedCircularDeps.includes(app?.id);
+            })
```

Comment:
It is actually not the same code. One is checking against north bound and one is checking against southbound.

Reply:
We could still factor out the common logic, if you want to. Something like:

```ts
  const filterOutCircularDeps = (depsToCheck: ApplicationDependency[]) =>
    (applications || []).filter((app) => {
      let blacklistedCircularDeps: (number | undefined)[] = [];
      if (depsToCheck) {
        depsToCheck?.forEach((dep) => {
          if (dep.from.id === application.id) {
            let matchingDep = dep;
            blacklistedCircularDeps = depsToCheck
              .filter((dep) => dep.from.id === matchingDep.to.id)
              .map((dep) => dep.to.id);
          }
        });
      }
      return !blacklistedCircularDeps.includes(app?.id);
    });
```

and then used here like:
```tsx
options={filterOutCircularDeps(southDependencies)}
```
or
```tsx
options={filterOutCircularDeps(northDependencies)
```

The logic is also a bit hard to read imo. you're doing a `forEach` with a condition that, if met, will reassign `blacklistedCircularDeps`. Do you expect only one dependency to match this condition? you could use a `.find()` maybe. not sure if it matters.

Reply:
> there can be multiple deps defined for a given apps nb and sb coordinates
I think this would result in not all of those getting used then, right? you only end up with the latest value of `blacklistedCircularDeps` at the end of the `forEach` and filter based on only that. maybe you do want the forEach and instead of this

```ts
            blacklistedCircularDeps = depsToCheck
              .filter((dep) => dep.from.id === matchingDep.to.id)
              .map((dep) => dep.to.id);
```

maybe you want this?

```ts
            blacklistedCircularDeps = [
              ...blacklistedCircularDeps,
              ...depsToCheck
                .filter((dep) => dep.from.id === matchingDep.to.id)
                .map((dep) => dep.to.id),
            ];
```

Reply:
Yeah, that's what I meant, factor the common logic.

Reply:
For the list of available southbound apps, we want to check that for the selected app there is no conflicting northbound dependency up the chain. I.e If I have apps 'core', 'north', and 'south, I want to make sure that if app 'south' is a downstream dependency of app 'core', I can't add app 'south' as a northbound dependency of app 'north'. Also, I want to make sure that I can't add app 'north' as a downstream dependency of app 'south' since app 'north' already has an existing conflicting mapping to app 'core' as an upstream dependency. 


```
 .filter((app) => {
        let blacklistedCircularDeps: (number | undefined)[] = [];
```
Case 1) Southbound dependency list - We are checking that there are no conflicting northbound dependencies.

For our current selected app ("application.id"), we want to find any dependency in the dependency list that has a dependency mapping "to" the current selected application. That would tell us that the current app has an existing northbound dependency. 
```
        if (dependencies && circularDependencyCheckDirection === "north") {
          const matchingDep = dependencies.find(
            (dep) => dep.to.id === application.id
          );
```

Now that we know that there is a northbound dependency, we can be sure to blacklist any application in the list that are  mapped as northbound of the application that is mapped as a northbound dependency of our current app. This can be done by creating a list of app IDs that meet the above scenario. 
```
          blacklistedCircularDeps = dependencies
            .filter((dep) => dep.to.id === matchingDep?.from.id)
            .map((dep) => dep.from.id);

```
The same can be done to check for conflicting southbound dependencies below. 
```
        } else if (
          dependencies &&
          circularDependencyCheckDirection === "south"
        ) {
          const matchingDep = dependencies.find(
            (dep) => dep.from.id === application.id
          );
          blacklistedCircularDeps = dependencies
            .filter((dep) => dep.from.id === matchingDep?.to.id)
            .map((dep) => dep.to.id);
        }
        return !blacklistedCircularDeps.includes(app?.id);
      })
```